File: examples/Python/03_visualizacion.py
"""
============================================
VISUALIZACIÓN DE DATOS EN PYTHON
============================================
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

# Configurar estilo
sns.set_style("whitegrid")
sns.set_palette("husl")

print("=" * 60)
print("VISUALIZACIÓN DE DATOS EN PYTHON")
print("=" * 60)

# ============================================
# 1. PREPARAR DATOS
# ============================================

np.random.seed(42)

# Diferentes tipos de datos
datos_normales = np.random.normal(75, 10, 200)
datos_sesgados = np.random.exponential(10, 200) + 40
datos_bimodal = np.concatenate([
    np.random.normal(60, 5, 100),
    np.random.normal(85, 5, 100)
])

# Datos para scatter plot
horas_estudio = np.random.uniform(1, 10, 50)
calificaciones = 40 + 5 * horas_estudio + np.random.normal(0, 5, 50)
calificaciones = np.clip(calificaciones, 0, 100)

# Datos categóricos
categorias = np.random.choice(['A', 'B', 'C', 'D'], size=100, 
                              p=[0.3, 0.25, 0.25, 0.2])

print("Datos generados exitosamente")

# ============================================
# 2. HISTOGRAMAS
# ============================================

print("\n" + "=" * 60)
print("CREANDO HISTOGRAMAS")
print("=" * 60)

fig, axes = plt.subplots(2, 2, figsize=(14, 10))
fig.suptitle('Análisis con Histogramas', fontsize=16, fontweight='bold')

# Histograma 1: Normal
axes[0, 0].hist(datos_normales, bins=25, color='lightblue', edgecolor='darkblue', alpha=0.7)
axes[0, 0].axvline(np.mean(datos_normales), color='red', linestyle='--', 
                   linewidth=2, label=f'Media = {np.mean(datos_normales):.1f}')
axes[0, 0].set_title('Distribución Normal')
axes[0, 0].set_xlabel('Valores')
axes[0, 0].set_ylabel('Frecuencia')
axes[0, 0].legend()
axes[0, 0].grid(True, alpha=0.3)

# Histograma 2: Sesgada
axes[0, 1].hist(datos_sesgados, bins=25, color='lightcoral', edgecolor='darkred', alpha=0.7)
axes[0, 1].axvline(np.mean(datos_sesgados), color='red', linestyle='--', linewidth=2)
axes[0, 1].axvline(np.median(datos_sesgados), color='green', linestyle='--', linewidth=2)
axes[0, 1].set_title('Distribución Sesgada')
axes[0, 1].set_xlabel('Valores')
axes[0, 1].set_ylabel('Frecuencia')
axes[0, 1].grid(True, alpha=0.3)

# Histograma 3: Bimodal
axes[1, 0].hist(datos_bimodal, bins=25, color='lightgreen', edgecolor='darkgreen', alpha=0.7)
axes[1, 0].set_title('Distribución Bimodal')
axes[1, 0].set_xlabel('Valores')
axes[1, 0].set_ylabel('Frecuencia')
axes[1, 0].grid(True, alpha=0.3)

# Histograma 4: Con densidad
axes[1, 1].hist(datos_normales, bins=30, density=True, color='lightyellow', 
                edgecolor='orange', alpha=0.7, label='Histograma')
# Agregar curva de densidad
from scipy.stats import gaussian_kde
density = gaussian_kde(datos_normales)
xs = np.linspace(datos_normales.min(), datos_normales.max(), 200)
axes[1, 1].plot(xs, density(xs), 'r-', linewidth=2, label='Densidad')
axes[1, 1].set_title('Histograma con Curva de Densidad')
axes[1, 1].set_xlabel('Valores')
axes[1, 1].set_ylabel('Densidad')
axes[1, 1].legend()
axes[1, 1].grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig('/tmp/histogramas_python.png', dpi=150, bbox_inches='tight')
print("Histogramas guardados en: /tmp/histogramas_python.png")

# ============================================
# 3. BOXPLOTS (DIAGRAMAS DE CAJA)
# ============================================

print("\n" + "=" * 60)
print("CREANDO BOXPLOTS")
print("=" * 60)

fig, axes = plt.subplots(1, 2, figsize=(14, 6))

# Boxplot simple
axes[0].boxplot(datos_normales, vert=True)
axes[0].set_ylabel('Valores')
axes[0].set_title('Boxplot Simple')
axes[0].grid(True, alpha=0.3, axis='y')

# Boxplot comparativo
datos_comp = [datos_normales, datos_sesgados, datos_bimodal]
bp = axes[1].boxplot(datos_comp, labels=['Normal', 'Sesgada', 'Bimodal'],
                     patch_artist=True)
# Colorear cajas
colors = ['lightblue', 'lightcoral', 'lightgreen']
for patch, color in zip(bp['boxes'], colors):
    patch.set_facecolor(color)
axes[1].set_ylabel('Valores')
axes[1].set_title('Comparación de Distribuciones')
axes[1].grid(True, alpha=0.3, axis='y')

plt.tight_layout()
plt.savefig('/tmp/boxplots_python.png', dpi=150, bbox_inches='tight')
print("Boxplots guardados en: /tmp/boxplots_python.png")

# ============================================
# 4. BOXPLOTS CON SEABORN
# ============================================

# Crear DataFrame
df_dist = pd.DataFrame({
    'valores': np.concatenate([datos_normales, datos_sesgados, datos_bimodal]),
    'tipo': ['Normal']*200 + ['Sesgada']*200 + ['Bimodal']*200
})

plt.figure(figsize=(12, 6))

plt.subplot(1, 2, 1)
sns.boxplot(x='tipo', y='valores', data=df_dist, palette='Set2')
plt.title('Boxplot con Seaborn')
plt.ylabel('Valores')
plt.xlabel('Tipo de Distribución')
plt.grid(True, alpha=0.3, axis='y')

plt.subplot(1, 2, 2)
sns.violinplot(x='tipo', y='valores', data=df_dist, palette='Set2')
plt.title('Violin Plot')
plt.ylabel('Valores')
plt.xlabel('Tipo de Distribución')
plt.grid(True, alpha=0.3, axis='y')

plt.tight_layout()
plt.savefig('/tmp/seaborn_boxplots.png', dpi=150, bbox_inches='tight')
print("Boxplots seaborn guardados en: /tmp/seaborn_boxplots.png")

# ============================================
# 5. GRÁFICOS DE DISPERSIÓN
# ============================================

print("\n" + "=" * 60)
print("CREANDO SCATTER PLOTS")
print("=" * 60)

fig, axes = plt.subplots(1, 2, figsize=(14, 6))

# Scatter plot básico
axes[0].scatter(horas_estudio, calificaciones, s=80, alpha=0.6, color='steelblue')
# Línea de tendencia
z = np.polyfit(horas_estudio, calificaciones, 1)
p = np.poly1d(z)
axes[0].plot(horas_estudio, p(horas_estudio), "r--", linewidth=2, label='Tendencia')
axes[0].set_xlabel('Horas de Estudio')
axes[0].set_ylabel('Calificación')
axes[0].set_title('Relación: Estudio vs Calificación')
axes[0].legend()
axes[0].grid(True, alpha=0.3)

# Calcular correlación
correlacion = np.corrcoef(horas_estudio, calificaciones)[0, 1]
axes[0].text(2, 95, f'r = {correlacion:.3f}', fontsize=12, 
             bbox=dict(boxstyle='round', facecolor='wheat', alpha=0.5))

# Scatter plot con categorías
categorias_rend = pd.cut(calificaciones, bins=[0, 60, 80, 100], 
                         labels=['Bajo', 'Medio', 'Alto'])
colors_rend = {'Bajo': 'red', 'Medio': 'yellow', 'Alto': 'green'}

for cat in ['Bajo', 'Medio', 'Alto']:
    mask = categorias_rend == cat
    axes[1].scatter(horas_estudio[mask], calificaciones[mask], 
                   s=80, alpha=0.6, color=colors_rend[cat], label=cat)

axes[1].set_xlabel('Horas de Estudio')
axes[1].set_ylabel('Calificación')
axes[1].set_title('Con Categorías de Rendimiento')
axes[1].legend()
axes[1].grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig('/tmp/scatter_plots.png', dpi=150, bbox_inches='tight')
print("Scatter plots guardados en: /tmp/scatter_plots.png")

# ============================================
# 6. SCATTER PLOT CON SEABORN
# ============================================

df_scatter = pd.DataFrame({
    'Horas de Estudio': horas_estudio,
    'Calificación': calificaciones
})

plt.figure(figsize=(10, 6))
sns.regplot(x='Horas de Estudio', y='Calificación', data=df_scatter,
            scatter_kws={'s': 80, 'alpha': 0.6}, color='steelblue')
plt.title(f'Relación entre Estudio y Calificación (r = {correlacion:.3f})')
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.savefig('/tmp/seaborn_scatter.png', dpi=150, bbox_inches='tight')
print("Scatter seaborn guardado en: /tmp/seaborn_scatter.png")

# ============================================
# 7. GRÁFICOS DE BARRAS
# ============================================

print("\n" + "=" * 60)
print("CREANDO GRÁFICOS DE BARRAS")
print("=" * 60)

# Contar frecuencias
valores_unicos, conteos = np.unique(categorias, return_counts=True)
porcentajes = (conteos / len(categorias)) * 100

fig, axes = plt.subplots(2, 2, figsize=(14, 10))

# Barras verticales
axes[0, 0].bar(valores_unicos, conteos, color=['#FF6B6B', '#4ECDC4', '#45B7D1', '#FFA07A'],
               edgecolor='black', alpha=0.7)
axes[0, 0].set_xlabel('Categoría')
axes[0, 0].set_ylabel('Frecuencia')
axes[0, 0].set_title('Distribución de Categorías')
axes[0, 0].grid(True, alpha=0.3, axis='y')

# Barras horizontales
axes[0, 1].barh(valores_unicos, conteos, color=['#FF6B6B', '#4ECDC4', '#45B7D1', '#FFA07A'],
                edgecolor='black', alpha=0.7)
axes[0, 1].set_ylabel('Categoría')
axes[0, 1].set_xlabel('Frecuencia')
axes[0, 1].set_title('Barras Horizontales')
axes[0, 1].grid(True, alpha=0.3, axis='x')

# Barras con porcentajes
bars = axes[1, 0].bar(valores_unicos, porcentajes, 
                      color=['#FF6B6B', '#4ECDC4', '#45B7D1', '#FFA07A'],
                      edgecolor='black', alpha=0.7)
axes[1, 0].set_xlabel('Categoría')
axes[1, 0].set_ylabel('Porcentaje (%)')
axes[1, 0].set_title('Distribución en Porcentaje')
axes[1, 0].grid(True, alpha=0.3, axis='y')

# Agregar etiquetas
for bar, pct in zip(bars, porcentajes):
    height = bar.get_height()
    axes[1, 0].text(bar.get_x() + bar.get_width()/2., height + 1,
                    f'{pct:.1f}%', ha='center', va='bottom', fontweight='bold')

# Gráfico de pastel
axes[1, 1].pie(conteos, labels=valores_unicos, autopct='%1.1f%%',
               colors=['#FF6B6B', '#4ECDC4', '#45B7D1', '#FFA07A'],
               startangle=90, explode=[0.05]*len(valores_unicos))
axes[1, 1].set_title('Gráfico de Pastel')

plt.tight_layout()
plt.savefig('/tmp/graficos_barras.png', dpi=150, bbox_inches='tight')
print("Gráficos de barras guardados en: /tmp/graficos_barras.png")

# ============================================
# 8. GRÁFICOS DE LÍNEAS (SERIES TEMPORALES)
# ============================================

print("\n" + "=" * 60)
print("CREANDO GRÁFICOS DE LÍNEAS")
print("=" * 60)

# Simular serie temporal
tiempo = np.arange(1, 101)
tendencia = 50 + 0.3 * tiempo
estacionalidad = 10 * np.sin(tiempo / 5)
ruido = np.random.normal(0, 3, 100)
serie_temporal = tendencia + estacionalidad + ruido

plt.figure(figsize=(14, 6))

plt.subplot(1, 2, 1)
plt.plot(tiempo, serie_temporal, 'b-', linewidth=2, label='Datos observados', alpha=0.7)
plt.plot(tiempo, tendencia, 'r--', linewidth=2, label='Tendencia')
plt.scatter(tiempo, serie_temporal, s=20, alpha=0.3, color='blue')
plt.xlabel('Tiempo')
plt.ylabel('Valor')
plt.title('Serie Temporal con Tendencia')
plt.legend()
plt.grid(True, alpha=0.3)

plt.subplot(1, 2, 2)
plt.plot(tiempo, estacionalidad, 'g-', linewidth=2, label='Componente estacional')
plt.plot(tiempo, ruido, 'orange', alpha=0.5, label='Ruido')
plt.xlabel('Tiempo')
plt.ylabel('Valor')
plt.title('Componentes de la Serie')
plt.legend()
plt.grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig('/tmp/series_temporales.png', dpi=150, bbox_inches='tight')
print("Series temporales guardadas en: /tmp/series_temporales.png")

# ============================================
# 9. HEATMAP (MATRIZ DE CORRELACIÓN)
# ============================================

print("\n" + "=" * 60)
print("CREANDO HEATMAP DE CORRELACIÓN")
print("=" * 60)

# Crear datos multivariados con correlaciones
np.random.seed(999)
n = 100
var1 = np.random.randn(n)
var2 = var1 * 0.8 + np.random.randn(n) * 0.4
var3 = -var1 * 0.6 + np.random.randn(n) * 0.5
var4 = np.random.randn(n)

df_multi = pd.DataFrame({
    'Variable 1': var1,
    'Variable 2': var2,
    'Variable 3': var3,
    'Variable 4': var4
})

# Calcular matriz de correlación
matriz_cor = df_multi.corr()

plt.figure(figsize=(10, 8))
sns.heatmap(matriz_cor, annot=True, fmt='.2f', cmap='coolwarm', 
            center=0, square=True, linewidths=1, cbar_kws={"shrink": 0.8})
plt.title('Matriz de Correlación', fontsize=14, fontweight='bold')
plt.tight_layout()
plt.savefig('/tmp/heatmap_correlacion.png', dpi=150, bbox_inches='tight')
print("Heatmap guardado en: /tmp/heatmap_correlacion.png")

# ============================================
# 10. PANEL COMBINADO
# ============================================

print("\n" + "=" * 60)
print("CREANDO PANEL COMBINADO")
print("=" * 60)

fig = plt.figure(figsize=(16, 12))
gs = fig.add_gridspec(3, 3, hspace=0.3, wspace=0.3)

# Histograma
ax1 = fig.add_subplot(gs[0, :2])
ax1.hist(datos_normales, bins=30, color='lightblue', edgecolor='darkblue', alpha=0.7)
ax1.set_title('Histograma', fontweight='bold')
ax1.set_xlabel('Valores')
ax1.set_ylabel('Frecuencia')
ax1.grid(True, alpha=0.3)

# Boxplot
ax2 = fig.add_subplot(gs[0, 2])
ax2.boxplot([datos_normales, datos_sesgados], labels=['Normal', 'Sesgada'])
ax2.set_title('Boxplots', fontweight='bold')
ax2.grid(True, alpha=0.3, axis='y')

# Scatter
ax3 = fig.add_subplot(gs[1, :2])
ax3.scatter(horas_estudio, calificaciones, s=60, alpha=0.6, color='steelblue')
ax3.plot(horas_estudio, p(horas_estudio), 'r--', linewidth=2)
ax3.set_title('Scatter Plot', fontweight='bold')
ax3.set_xlabel('Horas de Estudio')
ax3.set_ylabel('Calificación')
ax3.grid(True, alpha=0.3)

# Barras
ax4 = fig.add_subplot(gs[1, 2])
ax4.bar(valores_unicos, conteos, color=['#FF6B6B', '#4ECDC4', '#45B7D1', '#FFA07A'], alpha=0.7)
ax4.set_title('Gráfico de Barras', fontweight='bold')
ax4.grid(True, alpha=0.3, axis='y')

# Serie temporal
ax5 = fig.add_subplot(gs[2, :])
ax5.plot(tiempo, serie_temporal, 'b-', linewidth=2, alpha=0.7)
ax5.plot(tiempo, tendencia, 'r--', linewidth=2)
ax5.set_title('Serie Temporal', fontweight='bold')
ax5.set_xlabel('Tiempo')
ax5.set_ylabel('Valor')
ax5.grid(True, alpha=0.3)

fig.suptitle('Panel Integrado de Visualizaciones', fontsize=16, fontweight='bold')
plt.savefig('/tmp/panel_completo.png', dpi=150, bbox_inches='tight')
print("Panel completo guardado en: /tmp/panel_completo.png")

print("\n" + "=" * 60)
print("FIN DEL SCRIPT")
print("Todos los gráficos han sido generados exitosamente")
print("=" * 60)

# plt.show()  # Descomentar para mostrar gráficos interactivos

File: examples/R/03_visualizacion.R
# ============================================
# VISUALIZACIÓN DE DATOS EN R
# ============================================

# Cargar librerías
library(ggplot2)
library(gridExtra)  # Para múltiples gráficos
# install.packages("gridExtra")  # Descomentar si necesario

# ============================================
# 1. PREPARAR DATOS
# ============================================

set.seed(42)

# Generar diferentes tipos de datos
datos_normales <- rnorm(200, mean = 75, sd = 10)
datos_sesgados <- rexp(200, rate = 0.05) + 40
datos_bimodal <- c(rnorm(100, mean = 60, sd = 5), rnorm(100, mean = 85, sd = 5))

# Datos para scatter plot
horas_estudio <- runif(50, min = 1, max = 10)
calificaciones <- 40 + 5 * horas_estudio + rnorm(50, mean = 0, sd = 5)
calificaciones <- pmin(calificaciones, 100)  # Limitar a 100

# Datos categóricos
categorias <- sample(c("A", "B", "C", "D"), 100, replace = TRUE, 
                     prob = c(0.3, 0.25, 0.25, 0.2))

# ============================================
# 2. HISTOGRAMAS
# ============================================

print("=== CREANDO HISTOGRAMAS ===")

# Histograma básico
par(mfrow = c(2, 2))

hist(datos_normales,
     main = "Distribución Normal",
     xlab = "Valores",
     ylab = "Frecuencia",
     col = "lightblue",
     border = "darkblue",
     breaks = 20)

hist(datos_sesgados,
     main = "Distribución Sesgada",
     xlab = "Valores",
     ylab = "Frecuencia",
     col = "lightcoral",
     border = "darkred",
     breaks = 20)

hist(datos_bimodal,
     main = "Distribución Bimodal",
     xlab = "Valores",
     ylab = "Frecuencia",
     col = "lightgreen",
     border = "darkgreen",
     breaks = 20)

# Histograma con densidad
hist(datos_normales,
     main = "Histograma con Curva de Densidad",
     xlab = "Valores",
     ylab = "Densidad",
     col = "lightyellow",
     border = "orange",
     probability = TRUE,
     breaks = 20)
lines(density(datos_normales), col = "red", lwd = 2)

# ============================================
# 3. HISTOGRAMAS CON GGPLOT2
# ============================================

par(mfrow = c(1, 1))

# Crear dataframe
df_hist <- data.frame(
  valores = c(datos_normales, datos_sesgados, datos_bimodal),
  tipo = rep(c("Normal", "Sesgada", "Bimodal"), each = 200)
)

# Histogramas facetados
p1 <- ggplot(df_hist, aes(x = valores, fill = tipo)) +
  geom_histogram(bins = 25, color = "white", alpha = 0.7) +
  facet_wrap(~tipo, scales = "free") +
  labs(title = "Comparación de Distribuciones",
       x = "Valores",
       y = "Frecuencia") +
  theme_minimal() +
  theme(legend.position = "none")

print(p1)

# ============================================
# 4. BOXPLOTS (DIAGRAMAS DE CAJA)
# ============================================

print("\n=== CREANDO BOXPLOTS ===")

# Boxplot básico
par(mfrow = c(1, 2))

boxplot(datos_normales,
        main = "Boxplot Simple",
        ylab = "Valores",
        col = "lightblue",
        border = "darkblue",
        notch = FALSE)

# Agregar puntos de datos
stripchart(datos_normales, 
           vertical = TRUE, 
           method = "jitter",
           add = TRUE, 
           pch = 20, 
           col = rgb(0, 0, 1, 0.2))

# Boxplot comparativo
datos_grupos <- list(
  Normal = datos_normales,
  Sesgada = datos_sesgados,
  Bimodal = datos_bimodal
)

boxplot(datos_grupos,
        main = "Comparación de Distribuciones",
        ylab = "Valores",
        col = c("lightblue", "lightcoral", "lightgreen"),
        border = c("darkblue", "darkred", "darkgreen"),
        las = 1)

# ============================================
# 5. BOXPLOTS CON GGPLOT2
# ============================================

par(mfrow = c(1, 1))

p2 <- ggplot(df_hist, aes(x = tipo, y = valores, fill = tipo)) +
  geom_boxplot(alpha = 0.7, outlier.color = "red", outlier.size = 2) +
  geom_jitter(width = 0.2, alpha = 0.1) +
  labs(title = "Boxplots con Puntos de Datos",
       x = "Tipo de Distribución",
       y = "Valores") +
  theme_minimal() +
  theme(legend.position = "none") +
  scale_fill_brewer(palette = "Set2")

print(p2)

# ============================================
# 6. GRÁFICOS DE DISPERSIÓN (SCATTER PLOTS)
# ============================================

print("\n=== CREANDO SCATTER PLOTS ===")

# Scatter plot básico
par(mfrow = c(1, 2))

plot(horas_estudio, calificaciones,
     main = "Relación: Estudio vs Calificación",
     xlab = "Horas de Estudio",
     ylab = "Calificación",
     pch = 19,
     col = rgb(0, 0, 1, 0.5),
     cex = 1.5)

# Agregar línea de tendencia
modelo <- lm(calificaciones ~ horas_estudio)
abline(modelo, col = "red", lwd = 2)

# Agregar grid
grid()

# Correlación
correlacion <- cor(horas_estudio, calificaciones)
text(2, 95, paste("r =", round(correlacion, 3)), cex = 1.2)

# Scatter plot con categorías de tamaño
plot(horas_estudio, calificaciones,
     main = "Con Categorías de Rendimiento",
     xlab = "Horas de Estudio",
     ylab = "Calificación",
     pch = 21,
     bg = ifelse(calificaciones >= 80, "green", 
                 ifelse(calificaciones >= 60, "yellow", "red")),
     cex = 1.5)

legend("bottomright",
       legend = c("Alto (≥80)", "Medio (60-79)", "Bajo (<60)"),
       pch = 21,
       pt.bg = c("green", "yellow", "red"),
       cex = 0.8)

# ============================================
# 7. SCATTER PLOT CON GGPLOT2
# ============================================

par(mfrow = c(1, 1))

df_scatter <- data.frame(
  horas = horas_estudio,
  notas = calificaciones
)

p3 <- ggplot(df_scatter, aes(x = horas, y = notas)) +
  geom_point(alpha = 0.6, size = 3, color = "steelblue") +
  geom_smooth(method = "lm", color = "red", se = TRUE, fill = "pink") +
  labs(title = "Relación entre Horas de Estudio y Calificación",
       subtitle = paste("Correlación:", round(correlacion, 3)),
       x = "Horas de Estudio",
       y = "Calificación") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))

print(p3)

# ============================================
# 8. GRÁFICOS DE BARRAS
# ============================================

print("\n=== CREANDO GRÁFICOS DE BARRAS ===")

# Contar frecuencias
tabla_categorias <- table(categorias)

# Gráfico de barras básico
par(mfrow = c(2, 2))

barplot(tabla_categorias,
        main = "Distribución de Categorías",
        xlab = "Categoría",
        ylab = "Frecuencia",
        col = rainbow(length(tabla_categorias)),
        border = "white")

# Gráfico de barras horizontal
barplot(tabla_categorias,
        main = "Barras Horizontales",
        xlab = "Frecuencia",
        ylab = "Categoría",
        col = heat.colors(length(tabla_categorias)),
        border = "white",
        horiz = TRUE)

# Gráfico de barras con porcentajes
porcentajes <- prop.table(tabla_categorias) * 100
barplot(porcentajes,
        main = "Distribución en Porcentaje",
        xlab = "Categoría",
        ylab = "Porcentaje (%)",
        col = terrain.colors(length(porcentajes)),
        border = "white",
        ylim = c(0, max(porcentajes) * 1.2))

# Agregar etiquetas de porcentaje
text(x = seq_along(porcentajes) * 1.2 - 0.5,
     y = porcentajes + 2,
     labels = paste0(round(porcentajes, 1), "%"),
     cex = 0.8)

# Gráfico de pastel
pie(tabla_categorias,
    main = "Gráfico de Pastel",
    col = rainbow(length(tabla_categorias)),
    labels = paste(names(tabla_categorias), "\n", tabla_categorias, 
                   " (", round(porcentajes, 1), "%)", sep = ""))

# ============================================
# 9. GRÁFICOS DE BARRAS CON GGPLOT2
# ============================================

par(mfrow = c(1, 1))

df_barras <- data.frame(categoria = categorias)

p4 <- ggplot(df_barras, aes(x = categoria, fill = categoria)) +
  geom_bar() +
  geom_text(stat = 'count', aes(label = after_stat(count)), vjust = -0.5) +
  labs(title = "Distribución de Categorías",
       x = "Categoría",
       y = "Frecuencia") +
  theme_minimal() +
  scale_fill_brewer(palette = "Set3") +
  theme(legend.position = "none")

print(p4)

# ============================================
# 10. GRÁFICOS DE LÍNEAS (SERIES TEMPORALES)
# ============================================

print("\n=== CREANDO GRÁFICOS DE LÍNEAS ===")

# Simular serie temporal
tiempo <- 1:100
tendencia <- 50 + 0.3 * tiempo
estacionalidad <- 10 * sin(tiempo / 5)
ruido <- rnorm(100, mean = 0, sd = 3)
serie_temporal <- tendencia + estacionalidad + ruido

par(mfrow = c(1, 1))

plot(tiempo, serie_temporal,
     type = "l",
     main = "Serie Temporal con Tendencia",
     xlab = "Tiempo",
     ylab = "Valor",
     col = "blue",
     lwd = 2)

# Agregar línea de tendencia
lines(tiempo, tendencia, col = "red", lwd = 2, lty = 2)

# Agregar puntos
points(tiempo, serie_temporal, pch = 20, col = rgb(0, 0, 1, 0.3))

legend("topleft",
       legend = c("Datos observados", "Tendencia"),
       col = c("blue", "red"),
       lty = c(1, 2),
       lwd = 2)

grid()

# ============================================
# 11. GRÁFICOS DE DENSIDAD
# ============================================

print("\n=== CREANDO GRÁFICOS DE DENSIDAD ===")

par(mfrow = c(1, 1))

# Gráfico de densidad con ggplot2
p5 <- ggplot(df_hist, aes(x = valores, fill = tipo)) +
  geom_density(alpha = 0.5) +
  geom_vline(data = aggregate(valores ~ tipo, df_hist, mean),
             aes(xintercept = valores, color = tipo),
             linetype = "dashed", size = 1) +
  labs(title = "Curvas de Densidad Comparativas",
       subtitle = "Líneas verticales muestran la media",
       x = "Valores",
       y = "Densidad") +
  theme_minimal() +
  scale_fill_brewer(palette = "Set1")

print(p5)

# ============================================
# 12. GRÁFICOS COMBINADOS (PANEL)
# ============================================

print("\n=== CREANDO PANEL DE GRÁFICOS ===")

# Crear múltiples gráficos con ggplot2
p_panel1 <- ggplot(df_hist, aes(x = valores, fill = tipo)) +
  geom_histogram(bins = 20, alpha = 0.7) +
  facet_wrap(~tipo, scales = "free") +
  theme_minimal() +
  theme(legend.position = "none")

p_panel2 <- ggplot(df_hist, aes(x = tipo, y = valores, fill = tipo)) +
  geom_violin(alpha = 0.7) +
  theme_minimal() +
  theme(legend.position = "none")

p_panel3 <- ggplot(df_scatter, aes(x = horas, y = notas)) +
  geom_point(color = "steelblue", alpha = 0.6) +
  geom_smooth(method = "lm", color = "red") +
  theme_minimal()

p_panel4 <- ggplot(df_barras, aes(x = categoria, fill = categoria)) +
  geom_bar() +
  theme_minimal() +
  theme(legend.position = "none")

# Combinar en grid
grid.arrange(p_panel1, p_panel2, p_panel3, p_panel4, ncol = 2)

# ============================================
# 13. MATRIZ DE CORRELACIÓN (HEATMAP)
# ============================================

print("\n=== CREANDO HEATMAP DE CORRELACIÓN ===")

# Crear datos multivariados
set.seed(999)
n <- 100
datos_multi <- data.frame(
  var1 = rnorm(n),
  var2 = rnorm(n),
  var3 = rnorm(n),
  var4 = rnorm(n)
)

# Agregar correlaciones
datos_multi$var2 <- datos_multi$var1 * 0.8 + rnorm(n, sd = 0.4)
datos_multi$var3 <- -datos_multi$var1 * 0.6 + rnorm(n, sd = 0.5)

# Calcular matriz de correlación
matriz_cor <- cor(datos_multi)

# Heatmap básico
library(reshape2)
matriz_cor_melted <- melt(matriz_cor)

p6 <- ggplot(matriz_cor_melted, aes(Var1, Var2, fill = value)) +
  geom_tile(color = "white") +
  geom_text(aes(label = round(value, 2)), color = "black") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0, limit = c(-1, 1)) +
  labs(title = "Matriz de Correlación",
       x = "",
       y = "",
       fill = "Correlación") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

print(p6)

print("\n=== FIN DEL SCRIPT ===")
print("Todos los gráficos han sido generados exitosamente")
